Topic/some fine tuning

telephon opened the pull request:
Mainly: refactored the `*new` method, so that `That(\name) === That(\name)` and avoid double initialisation.

telephon committed:
no need to enforce callback

`nil.value` is a no-op.

also do not call prUpdateDefs twice

telephon committed:
cosmetics

telephon committed:
use instance if possible

telephon committed:
unit test for That(\name) access

telephon committed:
test for identity, not equality

capital-G commented on classes/that.sc:
```diff
@@ -21,51 +21,54 @@ That {
+	input_ { |newInput|
 		input = newInput;
-		this.prUpdateDefs();
+		this.prUpdateDefs;
```
I personally prefer to use `()` on function calls as I found this implicitness really confusing when I started sclang (especially considering that sclang also allows for higher order functions as well and all other languages I use don't just call the function in this case) and for me it leads to more readability b/c it removes one level of implicitness as it could just be a leftover of a parameter access.
It is also more consistent with the writing if the function needs parameters.

But at the same time, we can switch to this writing as well.

telephon replied:
Yes, I understand.

Nevertheless: Better switch, because this is idiomatic. 
One shouldn't think "function call" but message passing. A message with no arguments is just a word: 
```supercollider
{ (0..100).clump(3).scramble.flat }.dup.flop
```

capital-G replied:
I think if one is sc-native than this style of code is more common but especially on chained calls I find this extremely confusing as I can not separate between property and method. Also what happens if you need to pass a function as a parameter: does this get called as well (no, not yet, only when you call it) but the semantics help a lot in such cases to understand what one is passing around in such cases.

But as this is SC code one just has to accept it and start getting used to it I think.
For myself I really try to code as explicit as possible in SC as the implicitness of SC just took some of my lifetime already (fav of mine is `nil.size` is `0` combined with the fact that every unset variable is initiated as `nil` which includes not-set keys in an event).

capital-G commented on classes/that.sc:
```diff
@@ -21,51 +21,54 @@ That {
-				Error("Please provide an analyzerFunction").throw;
-			});
-			callback = callback ? {}; // enforce a function as callback and not allow nil
-			res = super.newCopyArgs(name, input, analyzerFunction, callback).init;
```
as we do not use `newCopyArgs` anymore: where do we attach the name property to the instance?

telephon replied:
`newCopyArgs` is just a convenience (and it can be more efficient if you initialise thousands of objects). We just set the variables via the `setFunctions` call, like any outsider could do.

capital-G commented on classes/that.sc:
```diff
@@ -21,51 +21,54 @@ That {
+		this.prUpdateDefs;
+	}
+
+	setFunctions { |newInput, newAnalyzerFunction, newCallback|
```
this should be a private function and should also be also excluded in the docs from showing up.

Also doesn't this crash if everything is not set/nil set during init?
I think we can default to an empty function `{}` for callback, but a `That` instance w/o an analyzer and input does not make sense to me and IMO it should be therefore impossible to create such a state as it leads to unintuitive interfaces? But maybe thats just my OCD from typesafety languages kicking in.

telephon replied:
> this should be a private function and should also be also excluded in the docs from showing up.

No, you want to be able to call this on an existing instance, too.

> Also doesn't this crash if everything is not set/nil set during init? I think we can default to an empty function `{}` for callback, 

It is typical in sclang to leave uninitialised values `nil`. The object `nil` responds to `value`, and one can always check with `notNil` to avoid building stuff. In proxies like `Ndef`, `nil` has also the meaning of a no-op and its always possible. The usage goes in the direction of a monoid.

>but a `That` instance w/o an analyzer and input does not make sense to me and IMO it should be therefore impossible to create such a state as it leads to unintuitive interfaces? But maybe thats just my OCD from typesafety languages kicking in.

The idea of a proxy like `That` is that you can use it before specifying it. So all the setters should be independent and atomic. You can e.g. also write `Ndef(\x, { Ndef(\y) + 2 })` without an `Ndef(\y)` in existence.

capital-G replied:
> No, you want to be able to call this on an existing instance, too.

Well, its a matter of taste. I am more a Python person with tendency to Java in such regards, citing from [the zen of Python](https://www.python.org/dev/peps/pep-0020/#toc-entry-2)

> * There should be one-- and preferably only one --obvious way to do it.

which conflicts with the current implementation as it allows already to set the input/analyzer function in 3 ways and now one needs to maintain and test all 3 ways. But I know that the SC does not follow this rule and is more in opposition to this rule but in my experience as a programmer I found this one of the most fruitful rules for a nice to use library.

Regarding
> The idea of a proxy like That is that you can use it before specifying it.

well, I see this differently as I don't understand the state of such an object and for me it just makes things more complicated as it introduces ambiguous states (or uninitialized state) to an object instead of enforcing a state which always is *valid*.

In the end it comes down to preference I think.

capital-G commented on classes/that.sc:
```diff
@@ -140,15 +144,15 @@ That {
 	}
 
-	*prCreateTrigger {|in, defaultTrigger, triggerFunction|
+	*prCreateTrigger { |in, defaultTrigger, triggerFunction|
```
did you do this manually or is there some linter for this? I used the SC IDE internal formatter and it did not fix this behavior.

telephon replied:
I do it manually, there is no linter because there is no syntax universally agreed on. So a matter of taste …

capital-G replied:
It would be maybe a good thing to enforce global styleguides via a linter as it avoids discussion about such stuff but regarding the C++ linter and the general ambiguity in SC I don't want to open a whole new cans of worms on this.
Just wanted to know if there is an easy way that I maybe did not know yet.

capital-G reviewed (changes requested):
Good point, but I am unsure about

* the lost(?) name parameter
* should we allow the state of unset input/analyzer or just throw an error (maybe at least give some hint via a warning/postln that an unintended state was reached?
* setFunctions should be a private function as there are already 2 ways to update the params of the class

edit: should we introduce a CHANGELOG.md so the changes get noted there?

telephon commented:
> Good point, but I am unsure about
> 
>     * the lost(?) name parameter

Ah I see, you keep the name in the instance. I'll fix that.


>     * should we allow the state of unset input/analyzer or just throw an error (maybe at least give some hint via a warning/postln that an unintended state was reached?

Can you explain what error that would be and where it would be thrown?

>     * setFunctions should be a private function as there are already 2 ways to update the params of the class

I think we need one way that changes them together because they are interdependent.
EDIT: otherwise, `prUpdateDefs` is called twice, once for each setter, instead of once for both together.

> edit: should we introduce a CHANGELOG.md so the changes get noted there?

Yes, once we are in alpha state ...

telephon committed:
keep name in instance

telephon committed:
only need to change if synth (input or analyzer) changed.

e.g. you could just write `That(\x, callback: { |e| e.postln })`. No need to touch the Ndef or OSCdef.

telephon committed:
fix both test and fix failed test
